src/blob_gc_job.cc

```diff
@@ -119,60 +119,68 @@ Status BlobGCJob::Run() {
     return s;
   }
 
-  std::string tmp;
-  for (const auto& f : blob_gc_->inputs()) {
-    if (!tmp.empty()) {
-      tmp.append(" ");
+  if (!blob_gc_->gc_sampled_inputs().empty()) {
+    Status s = DoRunGC();
+    if (!s.ok()) {
+      return s;
     }
-    tmp.append(std::to_string(f->file_number()));
   }
 
-  std::string tmp2;
-  for (const auto& f : blob_gc_->sampled_inputs()) {
-    if (!tmp2.empty()) {
-      tmp2.append(" ");
+  if (!blob_gc_->fs_sampled_inputs().empty()) {
+    Status s = DigHole();
+    if (!s.ok()) {
+      return s;
     }
-    tmp2.append(std::to_string(f->file_number()));
   }
 
-  ROCKS_LOG_BUFFER(log_buffer_, "[%s] Titan GC candidates[%s] selected[%s]",
-                   blob_gc_->column_family_handle()->GetName().c_str(),
-                   tmp.c_str(), tmp2.c_str());
-
-  if (blob_gc_->sampled_inputs().empty()) {
-    return Status::OK();
-  }
-
-  return DoRunGC();
+  return Status::OK();
 }
 
 Status BlobGCJob::SampleCandidateFiles() {
-  std::vector<BlobFileMeta*> result;
-  for (const auto& file : blob_gc_->inputs()) {
+  std::vector<BlobFileMeta*> gc_result;
+  std::vector<BlobFileMeta*> fs_result;
+  std::set<uint64_t> gc_selected_marks;
+
+  // select files for GC
+  for (const auto& file : blob_gc_->gc_inputs()) {
     bool selected = false;
-    Status s = DoSample(file, &selected);
-    if (!s.ok()) {
-      return s;
+    if (file->GetValidSize() <= blob_gc_->titan_cf_options().merge_small_file_threshold) {
+      selected = true;
+    } else {
+      Status s = DoSample(file, &selected);
+      if (!s.ok()) {
+        return s;
+      }
     }
     if (selected) {
-      result.push_back(file);
+      gc_selected_marks.insert(file->file_number());
+      gc_result.push_back(file);
     }
   }
-  if (!result.empty()) {
-    blob_gc_->set_sampled_inputs(std::move(result));
+
+  // select files for free space
+  for (const auto& file: blob_gc_->fs_inputs()) {
+    // don't select file that already be selected to GC
+    if (gc_selected_marks.find(file->file_number()) != gc_selected_marks.end()){
+      continue;
+    }
+
+    if (file->discardable_size() >= blob_gc_->titan_cf_options().free_space_threshold) {
+      fs_result.push_back(file);
+    }
+  }
+
+  if (!gc_result.empty()) {
```

Comment:
If gc_result is empty, can we sill call set_gc_sampled_inputs?

Reply:
well, I not sure since I am not too familiar with C++, but this can not be wrong since this code style come from pingcap/titan